plugins/SlicerT/SlicerT.cpp

```diff
@@ -0,0 +1,666 @@
+	0x0100,
+	Plugin::Type::Instrument,
+	new PluginPixmapLoader("icon"),
+	nullptr,
+	nullptr,
+};
+} // end extern
+
+PhaseVocoder::PhaseVocoder()
+	: m_FFTInput(s_windowSize, 0)
+	, m_IFFTReconstruction(s_windowSize, 0)
+	, m_allMagnitudes(s_windowSize, 0)
+	, m_allFrequencies(s_windowSize, 0)
+	, m_processedFreq(s_windowSize, 0)
+	, m_processedMagn(s_windowSize, 0)
+{
+	m_fftPlan = fftwf_plan_dft_r2c_1d(s_windowSize, m_FFTInput.data(), m_FFTSpectrum, FFTW_MEASURE);
+	m_ifftPlan = fftwf_plan_dft_c2r_1d(s_windowSize, m_FFTSpectrum, m_IFFTReconstruction.data(), FFTW_MEASURE);
+}
+
+PhaseVocoder::~PhaseVocoder()
+{
+	fftwf_destroy_plan(m_fftPlan);
+	fftwf_destroy_plan(m_ifftPlan);
+}
+
+void PhaseVocoder::loadData(std::vector<float> originalData, int sampleRate, float newRatio)
+{
+	m_dataLock.lock();
+
+	m_originalBuffer = originalData;
+	m_originalSampleRate = sampleRate;
+	m_scaleRatio = -1; // force update, kinda hacky
+
+	m_dataLock.unlock(); // stupid, but QRecursiveMutex is too expensive to have in updateParas and getFrames
+	updateParams(newRatio);
+	m_dataLock.lock();
+
+	// set buffer sizes
+	m_processedWindows.resize(m_numWindows, false);
+	m_lastPhase.resize(m_numWindows * s_windowSize, 0);
+	m_sumPhase.resize((m_numWindows + 1) * s_windowSize, 0);
+	m_freqCache.resize(m_numWindows * s_windowSize, 0);
+	m_magCache.resize(m_numWindows * s_windowSize, 0);
+
+	// clear phase buffers
+	std::fill(m_lastPhase.begin(), m_lastPhase.end(), 0);
+	std::fill(m_sumPhase.begin(), m_sumPhase.end(), 0);
+
+	// maybe limit this to a set amount of windows to reduce initial lag spikes
+	for (int i = 0; i < m_numWindows; i++)
+	{
+		if (!m_processedWindows[i])
+		{
+			generateWindow(i, false); // first pass, no cache
+			m_processedWindows[i] = true;
+		}
+	}
+
+	m_dataLock.unlock();
+}
+
+void PhaseVocoder::getFrames(std::vector<float>& outData, int start, int frames)
+{
+	if (m_originalBuffer.size() < 2048) { return; }
+	m_dataLock.lock();
+
+	if (m_scaleRatio == 1) { // directly copy original data
+		memcpy(outData.data(), m_originalBuffer.data() + start, frames*sizeof(float));
+		m_dataLock.unlock();
+		return;
+	}
+
+	int windowMargin = s_overSampling / 2; // numbers of windows before full quality
+	int startWindow = std::max(0.0f, (float)start / m_outStepSize - windowMargin);
+	int endWindow = std::min((float)m_numWindows, (float)(start + frames) / m_outStepSize + windowMargin);
+
+	// discard previous phaseSum if not processed
+	if (!m_processedWindows[startWindow])
+	{
+		std::fill_n(m_sumPhase.data() + startWindow * s_windowSize, s_windowSize, 0);
+	}
+
+	// this encompases the minimum windows needed to get full quality,
+	// which must be computed
+	for (int i = startWindow; i < endWindow; i++)
+	{
+		if (!m_processedWindows[i])
+		{
+			generateWindow(i, true); // theses should use the cache
+			m_processedWindows[i] = true;
+		}
+	}
+
+	for (int i = 0; i < frames; i++)
+	{
+		outData[i] = m_processedBuffer[start + i];
+	}
```

Comment:
Ranged-based for loop?

Reply:
How would this look with a range-based for loop? How does it handle the `start` offset?

Reply:
> How would this look with a range-based for loop? How does it handle the start offset?

Apologies. Yeah, it's probably better to use a regular for-loop here then. I thought we were just iterating it over a collection and trying to access its variables. No action needed here.